File: ShootR/ShootR/Server/Abilities/Boost.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace ShootR
{
    public class Boost : MovementAbility
    {
        // Increase speed by X, multiplies it
        public const double SPEED_INCREASE = 3;
        public const string NAME = "Boost";
        public static readonly TimeSpan DURATION = TimeSpan.FromSeconds(1.5);

        private ValueRef<bool> _controllable;
        
        public Boost(ShipMovementController movementController, ValueRef<bool> controllable)
            : base(NAME, movementController)
        {
            _controllable = controllable;
        }

        public override void Activate()
        {
            ShipMovementController smc = _movementController as ShipMovementController;
            smc.StopMovement();
            smc.Moving.Forward = true;

            _controllable.Value = false;
            MultiplySpeedBy(SPEED_INCREASE);
            base.Activate();
        }

        public override void Deactivate()
        {
            (_movementController as ShipMovementController).Moving.Forward = false;
            _controllable.Value = true;
            ResetSpeed();
            base.Deactivate();
        }

        public override void Update(DateTime utcNow)
        {
            if (Active && utcNow - ActivatedAt >= DURATION)
            {
                Deactivate();
                base.Update(utcNow);
            }
        }
    }
}

File: ShootR/ShootR/Client/Ships/Abilities/Boost.ts
/// <reference path="../../../Scripts/endgate-0.2.0.d.ts" />
/// <reference path="MovementAbility.ts" />
/// <reference path="../ShipMovementController.ts" />

module ShootR {

    export class Boost extends MovementAbility {
        public static NAME: string = "Boost";
        public static SPEED_INCREASE: number = 3; // Updated by server configuration value
        public static DURATION: eg.TimeSpan = eg.TimeSpan.FromSeconds(3); // Updated by server configuration value

        constructor(private _movementController: ShipMovementController) {
            super(Boost.NAME, _movementController);

            this.OnStart = new eg.EventHandler();
            this.OnStop = new eg.EventHandler();
        }

        public OnStart: eg.EventHandler;
        public OnStop: eg.EventHandler;

        public Activate(): void {
            this._movementController.StopAllMovement();
            this._movementController.Moving.Forward = true;            
            this._movementController.Controllable = false;

            if (!this.Active) {
                this.MultiplySpeedBy(Boost.SPEED_INCREASE);
                this.OnStart.Trigger();
            }

            super.Activate();
        }

        public Deactivate(): void {
            if (this.Active) {
                this.ResetSpeed();
                super.Deactivate();
                this._movementController.Moving.Forward = false;
                this._movementController.Controllable = true;
                this.OnStop.Trigger();
            }
        }

        public Update(gameTime: eg.GameTime): void {
            if (this.Active && eg.TimeSpan.DateSpan(this.ActivatedAt, gameTime.Now).Milliseconds >= Boost.DURATION.Milliseconds) {
                this.Deactivate();
            }
        }
    }
}